### ISSUE TITLE
SCSS error when importing a react shared-ui lib component into Next.js app

### ISSUE THREAD

**Issue by stopyransky** [nrwl/nx#7203]
## Current Behavior

i use shared-ui library with with react components, each shared component uses scss module and imports it in own tsx file. 
I export the component in library’s index.ts with `export * from './lib/mycomponent'`

When i import such library inside next.js app in nx monorepo setup I get "CSS Modules cannot be imported from within node_modules." error (log below)

I added next-transpile-modules plugin  to the next.config.js but it fails to find the shared ui lib.
I also tried to add transpileModules: [/^\@my-repo-name\/.*/],
in the next.config but it silently fails….
any ideas how i can transpile the scss files in shared ui library? do i need to change something in nx.json ?

## Expected Behavior
SCSS error is not present and shared-ui lib can be used in other apps

## Steps to Reproduce
example that replicates the error: https://github.com/stopyransky/nx-test-scss

### Failure Logs
```
error - ../../libs/shared-ui/src/lib/Spinner/spinner.module.scss
CSS Modules cannot be imported from within node_modules.
```

### Environment
  Node : 14.17.6
  OS   : darwin x64
  npm  : 6.14.15
  
  nx : Not Found
  @nrwl/angular : Not Found
  @nrwl/cli : 12.9.0
  @nrwl/cypress : 12.9.0
  @nrwl/devkit : 12.9.0
  @nrwl/eslint-plugin-nx : 12.9.0
  @nrwl/express : Not Found
  @nrwl/jest : 12.9.0
  @nrwl/linter : 12.9.0
  @nrwl/nest : Not Found
  @nrwl/next : 12.9.0
  @nrwl/node : Not Found
  @nrwl/nx-cloud : Not Found
  @nrwl/react : 12.9.0
  @nrwl/schematics : Not Found
  @nrwl/tao : 12.9.0
  @nrwl/web : 12.9.0
  @nrwl/workspace : 12.9.0
  @nrwl/storybook : 12.9.0
  @nrwl/gatsby : Not Found
  typescript : 4.3.5

**Comment by kirjai** [comment]
hi @stopyransky , thanks for opening the issue!

After checking out your repo, i can see that the `withNx` plugin that's supposed to be used in `next.config.js` files isn't actually being invoked in your case. so looks to me that perhaps you might be using the `next-compose-plugins` module wrong

could you double check that, to see if it fixes this issue?

**Comment by stopyransky** [comment]
Ok let me give some more verbose info why next.config.js is set up like this.

I need `next-transpile-modules` due to react-spring usage. Without it i get following error when running build: 

`nx build app --prod --verbose --dry-run --skip-nx-cache` causes following:

```bash
Unexpected token 'export'
info  - Collecting page data ./Users/.../node_modules/
@babel/runtime/helpers/esm/objectWithoutPropertiesLoose.js:1
export default function _objectWithoutPropertiesLoose(source, excluded) {
^^^^^^

SyntaxError: Unexpected token 'export'
    at Module._compile (internal/modules/cjs/loader.js:892:18)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Module.require (internal/modules/cjs/loader.js:849:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/Users/.../node_modules/react-spring/renderprops.js:7:53)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
```

So I need to add next-transpile-module (TM) and import react-spring modules from with cjs at the end (explained [**here**](https://github.com/pmndrs/react-spring/issues/601)), so `import { Spring } from "react-spring/renderprops.cjs"`.

If you refer to wrong setup in `next.config.js` because of this syntax:

`module.exports = withTM(withPlugins([...plugins], nextConfig));`

I have to do it this way because otherwise (unknown reason) the withTM will cause error on build.

Here is the 'proper' `next.config.js` withPlugins and withTM, that will fail with build error:

```js
const withNx = require('@nrwl/next/plugins/with-nx');
const withPlugins = require('next-compose-plugins');

const withTM = require('next-transpile-modules')(['react-spring'], {
  debug: true,
});

const withPWA = require('next-pwa');
const nextBaseConfig = require('../../next.base.config');

const nextConfig = {
  ...nextBaseConfig,
  reactStrictMode: true,
};

const plugins = [
  [withNx],
  [
    withPWA,
    {
      pwa: {
        dest: 'public',
        disable: process.env.NODE_ENV === 'development',
        importScripts: ['/workers/test-worker.js'],
      },
    },
  ],
  [withTM],
];
module.exports = withPlugins([...plugins], nextConfig);
```

resulting error: 

```bash
> nx run app:build:production --dryRun --verbose --skip-nx-cache
info  - Loaded env from apps/app/.env.local
info  - Using webpack 5. Reason: Enabled by default https://nextjs.org/docs/messages/webpack5
next-transpile-modules - trying to resolve the following modules:
  - react-spring
next-transpile-modules - the following paths will get transpiled:
  - /Users/.../node_modules/react-spring
info  - Using webpack 5. Reason: Enabled by default https://nextjs.org/docs/messages/webpack5
warn  - You have enabled experimental feature(s).
warn  - Experimental features are not covered by semver, and may cause unexpected or broken application behavior. Use them at your own risk.

next-transpile-modules - trying to resolve the following modules:
  - react-spring
next-transpile-modules - the following paths will get transpiled:
  - /Users/.../node_modules/react-spring
info  - Checking validity of types  
info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
> [PWA] Compile server
> [PWA] Compile client (static)
> [PWA] Auto register service worker with: /Users/.../node_modules/next-pwa/register.js
> [PWA] Service worker: /Users/.../apps/.../public/sw.js
> [PWA]   url: /sw.js
> [PWA]   scope: /
info  - Creating an optimized production build .next-transpile-modules - transpiled: /Users/.../node_modules/react-spring/renderprops.cjs.js
next-transpile-modules - transpiled: /Users/.../node_modules/react-spring/web.js
next-transpile-modules - transpiled: /Users/.../node_modules/react-spring/web.js
info  - Using external babel configuration from /Users/.../apps/.../.babelrc
(node:59936) [DEP_WEBPACK_MODULE_ISSUER] DeprecationWarning: Module.issuer: Use new ModuleGraph API
info  - Creating an optimized production build  
Failed to compile.

./pages/index.module.scss
CSS Modules cannot be imported from within node_modules.
Read more: https://nextjs.org/docs/messages/css-modules-npm
Location: pages/index.tsx

> Build failed because of webpack errors
Error: > Build failed because of webpack errors
    at /Users/.../node_modules/next/dist/build/index.js:390:19
    at async Span.traceAsyncFn (/Users/.../node_modules/next/dist/telemetry/trace/trace.js:60:20)
```

Here is the fancy next.config.js that will build successfully and SCSS module inside app will work:

```js
const withNx = require('@nrwl/next/plugins/with-nx');
const withPlugins = require('next-compose-plugins');

const withTM = require('next-transpile-modules')(['react-spring'], {
  debug: true,
});

const withPWA = require('next-pwa');
const nextBaseConfig = require('../../next.base.config');

const nextConfig = {
  ...nextBaseConfig,
  reactStrictMode: true,
};

const plugins = [
  [withNx],
  [
    withPWA,
    {
      pwa: {
        dest: 'public',
        disable: process.env.NODE_ENV === 'development',
        importScripts: ['/workers/test-worker.js'],
      },
    },
  ],
];

module.exports = withTM(withPlugins([...plugins], nextConfig));
``` 

Now we are on the same page.

If now I add to the app a shared-ui component that imports own SCSS style, it will break with following error: 


```bash
> nx run app:build:production --dryRun --verbose 
info  - Loaded env from apps/app/.env.local
info  - Using webpack 5. Reason: Enabled by default https://nextjs.org/docs/messages/webpack5
next-transpile-modules - trying to resolve the following modules:
  - react-spring
next-transpile-modules - the following paths will get transpiled:
  - /Users/.../node_modules/react-spring
info  - Using webpack 5. Reason: Enabled by default https://nextjs.org/docs/messages/webpack5
info  - Checking validity of types  
info  - Creating an optimized production build .next-transpile-modules - transpiled: /Users/.../node_modules/react-spring/renderprops.cjs.js
next-transpile-modules - transpiled: /Users/.../node_modules/react-spring/web.js
next-transpile-modules - transpiled: /Users/.../node_modules/react-spring/web.js
info  - Using external babel configuration from /Users/.../apps/app/.babelrc
info  - Using external babel configuration from /Users/.../apps/app/.babelrc
info  - Using external babel configuration from /Users/.../apps/app/.babelrc
(node:60723) [DEP_WEBPACK_MODULE_ISSUER] DeprecationWarning: Module.issuer: Use new ModuleGraph API
info  - Creating an optimized production build  
Failed to compile.

../../libs/shared-ui/src/lib/Spinner/spinner.module.scss
CSS Modules cannot be imported from within node_modules.
Read more: https://nextjs.org/docs/messages/css-modules-npm
Location: ../../libs/shared-ui/src/lib/Spinner/index.tsx

> Build failed because of webpack errors
Error: > Build failed because of webpack errors
    at /Users/.../node_modules/next/dist/build/index.js:390:19
    at async Span.traceAsyncFn (/Users/.../node_modules/next/dist/telemetry/trace/trace.js:60:20)
```

For me it seems that we have some combo error: either `next-transpile-modules` is broken (fancy config hack), or it is the 'withPlugins' or it is shared-ui SCSS import error - because hacky config works without shared-ui imports.

Any ideas how to fix this?  I would love to remove next-transpile-modules plugin but it is required by core dependency in the project.

**Comment by Raj-5497** [comment]
@kirjai  Any update about this??

**Comment by kirjai** [comment]
i've been coming back to this issue occasionally, but haven't gotten it to a working state, sadly. the cause is still the same - changes made to the webpack config by `next-transpile-modules` and by Nx aren't really compatible at the moment, sadly

**Comment by RayVR** [comment]
I am attempting to run nextjs + react native web in my nx workspace with shared libraries for a react native app.

If I use vanilla react-native directly in my next.js application everything works. The moment I try to import a library created using @nrwl:react or @nrwl/react-native, things fall apart.

As soon as I attempt to import code from a shared library I hit this error:
```
info  - Using external babel configuration from /home/<user>/Projects/<ws-name>/apps/braced-web/babel.config.js
error - ../../node_modules/react-native/Libraries/Utilities/binaryToBase64.js
Module parse failed: Unexpected token (15:28)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
| const base64 = require('base64-js');
|
> function binaryToBase64(data: ArrayBuffer | $ArrayBufferView): string {
|   if (data instanceof ArrayBuffer) {
|     // $FlowFixMe[reassign-const]
[ ready ] on http://localhost:4200

```

It seems like the correct approach would be to transpile the modules but if that is the case this issue with nx and next-transpile-modules would prevent that.

Is there any workaround at all to support library code that isn't just basic js/typescript?

**Comment by jaysoo** [comment]
The problem with `next-transpile-module` is that it breaks Nx's webpack configuration.

The workaround for now is to apply `withNx` first, then `withTM`. It's a bit counter intuitive, but calling `withTM` after `withNX` applies their webpack config _before_ Nx's.

Since both `withTM` and `withNx` modify similar webpack loaders, the former causes issues when used with Nx.

---

**TL;DR:** Your `next.config.js` needs to call `withNx` first, then `withTM`. I'll make a patch to make this ordering not necessary.

**Comment by ALexander4295502** [comment]
Not sure if this is related, I am still encountering the error 
```
Failed to compile
./pages/index.module.scss
CSS Modules cannot be imported from within node_modules.
Read more: https://nextjs.org/docs/messages/css-modules-npm
Location: pages/index.tsx
```
when I try to import an scss module from my local nx app (not from node_modules). 

the `next.config.js` on my side is:
```javascript
//@ts-check

// eslint-disable-next-line @typescript-eslint/no-var-requires
const withTM = require('next-transpile-modules')(['@heartlandone/vega-react']);
const { withNx } = require('@nrwl/next/plugins/with-nx');

/**
 * @type {import('@nrwl/next/plugins/with-nx').WithNxOptions}
 **/
const nextConfig = {
  nx: {
    // Set this to true if you would like to to use SVGR
    // See: https://github.com/gregberge/svgr
    svgr: false,
  },
};

module.exports = withTM(withNx(nextConfig));
```
 
I raised an issue in https://github.com/martpie/next-transpile-modules/issues/275 as well

**Comment by zhaoguangyue** [comment]
I have a similar problem

**Comment by jaysoo** [comment]
This should be fixed with the latest Nx. https://github.com/jaysoo/nx-issue-7203

Note: If you are using `next-transpile-module` please remove it since it is built into Next.js 13.1 and can cause issues if used.
